### PR TITLE
Optuna

### PR DESCRIPTION
**Pull request by viktorvaladi**
Adding new generic setup for attack objects to use optuna.

An attack object needs three functions to use this method:

reset_attack() # resets an attack to its initial state
suggest_parameters(trial) # gives a range and specifies which parameters to optimize
get_configs() # get attack config

Also this PR abstract out some functionality from GIA's to reduce code duplication.

### CODE REVIEW

File: leakpro/optuna.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                # save results if not pruned
+                if result_object is not None:
+                    result_object.save(name="optuna", path="./leakpro_output/results", config=attack_object.get_configs())
+                    return intermediary_results
+        elif isinstance(result, MIAResult):
+            return result.accuracy # add something reasonable to optimize toward here
+        return None
+
+    # Define the pruner and study
+    pruner = optuna.pruners.MedianPruner(n_warmup_steps=n_warmup_steps)
```
Comment (johanos1) [inline leakpro/optuna.py:39]: allow for different pruners and samplers on the input?

File: leakpro/optuna.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+from torch import Tensor
+
+from leakpro.attacks.gia_attacks.abstract_gia import AbstractGIA
+from leakpro.attacks.mia_attacks.abstract_mia import AbstractMIA
+from leakpro.metrics.attack_result import MIAResult
+from leakpro.utils.logger import logger
+from leakpro.utils.seed import seed_everything
+
+
+def optuna_optimal_hyperparameters(attack_object: Union[AbstractGIA, AbstractMIA], n_warmup_steps:int= 5,
```
Comment (johanos1) [inline leakpro/optuna.py:15]: Make an assess to ensure attack_object is the correct type.

File: pyproject.toml [hunk 3]
```diff
@@ -20,6 +20,9 @@ dependencies = [
   "joblib",
   "pyyaml",
   "scikit-learn",
+  "pytorch-ignite",
```
Comment (johanos1) [inline pyproject.toml:23]: ignite should only be in `dev` and `federated`

File: pyproject.toml [hunk 4]
```diff
@@ -20,6 +20,9 @@ dependencies = [
...
+  "gdown",
+  "optuna",
```
Comment (johanos1) [inline pyproject.toml:25]: optuna should not be in synthetic data as of now

File: pyproject.toml [hunk 5]
```diff
@@ -20,6 +20,9 @@ dependencies = [
   "joblib",
   "pyyaml",
   "scikit-learn",
+  "pytorch-ignite",
+  "gdown",
```
Comment (johanos1) [inline pyproject.toml:24]: why do we need gdown here? Is it for the examples? If so, it should be specified in there and not be installed in Leakpro.

File: leakpro/fl_utils/similarity_measurements.py [hunk 6]
```diff
@@ -1,13 +1,38 @@
...
 from ignite.metrics import SSIM
-from torch import Tensor, abs, cuda, mean, nn, no_grad, norm
-from torch.nn.modules.utils import _pair, _quadruple
+from torch import Tensor, abs, cuda, mean, no_grad, norm
 from torch.utils.data import DataLoader
 from torchmetrics.functional import peak_signal_noise_ratio
 
-from leakpro.utils.import_helper import Self
 
+def cosine_similarity_weights(client_gradient: torch.Tensor, reconstruction_gradient: torch.Tensor,
```
Comment (johanos1) [inline leakpro/fl_utils/similarity_measurements.py:9]: Nice. So now both inverting and Huang use the same function. Did you test that both still work?

File: leakpro/attacks/gia_attacks/invertinggradients.py [hunk 7]
```diff
@@ -157,42 +126,34 @@ def closure() -> torch.Tensor:
...
-            filtered_input_gradients = [client_gradient[i] for i in indices]
-        costs = sum((x * y).sum() for x, y in zip(filtered_input_gradients,
-                                                  filtered_trial_gradients))
-
-        trial_norm = sum(x.pow(2).sum()
-                         for x in filtered_trial_gradients).sqrt()
-        input_norm = sum(y.pow(2).sum()
-                         for y in filtered_input_gradients).sqrt()
-        return 1 - (costs / trial_norm / input_norm)
+    def reset_attack(self: Self) -> None:
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/invertinggradients.py:148]: Can we make it so this is called in the constructor? It would make sense to reset when creating the object?

File: leakpro/attacks/gia_attacks/huang.py [hunk 8]
```diff
@@ -172,46 +142,43 @@ def closure() -> torch.Tensor:
...
-            filtered_input_gradients = [client_gradient[i] for i in indices]
-        costs = sum((x * y).sum() for x, y in zip(filtered_input_gradients,
-                                                  filtered_trial_gradients))
-
-        trial_norm = sum(x.pow(2).sum()
-                         for x in filtered_trial_gradients).sqrt()
-        input_norm = sum(y.pow(2).sum()
-                         for y in filtered_input_gradients).sqrt()
-        return 1 - (costs / trial_norm / input_norm)
+    def reset_attack(self: Self) -> None:
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/huang.py:173]: Could we call this also in the constructor

File: leakpro/attacks/gia_attacks/invertinggradients.py [hunk 9]
```diff
@@ -157,42 +126,34 @@ def closure() -> torch.Tensor:
...
-        This function calculates the pairwise costs between each client gradient and the reconstruction gradient
-        using the cosine similarity measure. The costs are accumulated and averaged over all client gradients.
+    def _configure_attack(self: Self, configs: dict) -> None:
+        pass
 
-        Returns
-        -------
-            torch.Tensor: The average reconstruction cost.
+    def suggest_parameters(self: Self, trial: optuna.trial.Trial) -> None:
+        """Suggest parameters to chose and range for optimization for the Huang attack."""
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/invertinggradients.py:144]: Replace "Huang attack" with Inverting Gradient

File: leakpro/attacks/gia_attacks/huang.py [hunk 10]
```diff
@@ -172,46 +142,43 @@ def closure() -> torch.Tensor:
...
             loss_r_feature = sum([
                 mod.r_feature
                 for (idx, mod) in enumerate(self.loss_r_feature_layers)
             ])
 
             # Add the TV loss term to penalize large variations between pixels, encouraging smoother images.
-            rec_loss += self.t_v_scale * total_variation(self.reconstruction)
-            rec_loss += self.bn_reg * loss_r_feature
-            rec_loss += self.l2_scale * l2_norm(self.reconstruction)
+            rec_loss += self.configs.total_variation * total_variation(self.reconstruction)
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/huang.py:155]: I thought this was a bool. Please rename to something with scale or reg

**Review by johanos1 — CHANGES_REQUESTED** [review]

**Comment by johanos1** [comment]
Please also link the relevant issues in the PR.

File: examples/gia/huang_optuna/main.py [hunk 11]
```diff
@@ -2,16 +2,16 @@
 
 from cifar import get_cifar10_loader
 
+from leakpro.attacks.gia_attacks.huang import Huang, HuangConfig
 from leakpro.fl_utils.gia_train import train
```
[the remark this answers is not in the document]
Comment (viktorvaladi) [reply examples/gia/huang_optuna/main.py:6]: There's no run function here, the optuna method takes an attack object as input

File: examples/gia/huang_optuna/main.py [hunk 12]
```diff
@@ -2,16 +2,16 @@
 
 from cifar import get_cifar10_loader
 
+from leakpro.attacks.gia_attacks.huang import Huang, HuangConfig
```
[the remark this answers is not in the document]
Comment (viktorvaladi) [reply examples/gia/huang_optuna/main.py:5]: I think exposing the object to users is a decent level of abstraction, to allow for things like overriding functionality etc. I'm thinking of moving away from the "run methods" for GIA, to instead have the default being to expose the attack objects. 

But maybe it's okay to have leakpro.run("huang") as a lazy option as well, could investigate it

File: leakpro/attacks/gia_attacks/abstract_gia.py [hunk 13]
```diff
@@ -57,17 +64,66 @@ def prepare_attack(self:Self) -> None:
...
+                    )
+                if (i +1) % 500 == 0 and median_pooling:
+                    reconstruction.data = MedianPool2d(kernel_size=3, stride=1, padding=1, same=False)(reconstruction)
+            # Chose image who has given least loss
+            if loss < self.best_loss:
+                self.best_loss = loss
+                self.best_reconstruction = deepcopy(reconstruction_loader)
+                self.best_reconstruction_round = i
+                logger.info(f"New best loss: {loss} on round: {i}")
+            if i % 250 == 0:
```
[the remark this answers is not in the document]
Comment (viktorvaladi) [reply leakpro/attacks/gia_attacks/abstract_gia.py:121]: Yes exactly

File: leakpro/optuna.py (hunk 1 above)
Comment (viktorvaladi) [reply leakpro/optuna.py:39]: I think if we do that then we should allow for different optuna.pruners, optuna.samplers and optuna.storage. But I'm not sure how they work with e.g. hyperparameters etc - i think this can be future work for when we have more examples with optuna and know what we might want to vary?

File: leakpro/fl_utils/similarity_measurements.py (hunk 6 above)
Comment (viktorvaladi) [reply leakpro/fl_utils/similarity_measurements.py:9]: Yep works for both of them. I really don't understand why though. But whatever works works i guess

File: leakpro/attacks/gia_attacks/invertinggradients.py (hunk 7 above)
Comment (viktorvaladi) [reply leakpro/attacks/gia_attacks/invertinggradients.py:148]: Would it? It would already be in its initial state so it would just call the same things twice right?

File: leakpro/attacks/gia_attacks/huang.py (hunk 8 above)
Comment (viktorvaladi) [reply leakpro/attacks/gia_attacks/huang.py:173]: as above

File: examples/gia/huang_optuna/main.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (johanos1) [reply examples/gia/huang_optuna/main.py:6]: Sure, but doesnt it make more sense to just send `base_model, cifar10_loader, mean, std` into Leakpro. That way you can use the same interface with other attacks. Then you can have a flag to decide to use Optuna? Also a string for the attack?

File: examples/gia/huang_optuna/main.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (johanos1) [reply examples/gia/huang_optuna/main.py:5]: What is the benefit of exposing it like this? I mean, why would you want to override the functionality of Huang?

File: leakpro/optuna.py (hunk 1 above)
Comment (johanos1) [reply leakpro/optuna.py]: fair enough.

File: leakpro/fl_utils/similarity_measurements.py (hunk 6 above)
Comment (johanos1) [reply leakpro/fl_utils/similarity_measurements.py:9]: I guess the sign differ in Geiping and it therefore works with a max..

**Review by johanos1 — APPROVED** [review]
I like the update!
